### PR TITLE
Fix issue with Sigma mint creation

### PR DESCRIPTION
**Pull request by riordant**
## PR intention
- if using the same `wallet.dat` between two machines, the count is not synced. Therefore during mint transaction creation, we have to:
    - check the chain to see if this mint already exists
    - ensure the chain is synced for access to all mints

This PR adds that.

- Adds tests for the same

- reverts a bad change in `ResetMintCount` - The intention of this function is to reset the internal mint count to the value in the database.

- Adds a function `GetKeyFromKeypath` which returns a CPubKey given the parameters of a HD keypath. This allows pulling an existing key during `CreateMintSeed` if the key already exists

### CODE REVIEW

File: src/hdmint/wallet.h [hunk 1]
```diff
@@ -32,7 +32,8 @@ class CHDMintWallet
 
     bool SetupWallet(const uint160& hashSeedMaster, bool fResetCount=false);
     void SyncWithChain(bool fGenerateMintPool = true, boost::optional<std::list<std::pair<uint256, MintPoolEntry>>> listMints = boost::none);
-    bool GenerateMint(const sigma::CoinDenomination denom, sigma::PrivateCoin& coin, CHDMint& dMint, boost::optional<MintPoolEntry> mintPoolEntry = boost::none);
+    bool GetHDMintFromMintPoolEntry(const sigma::CoinDenomination& denom, sigma::PrivateCoin& coin, CHDMint& dMint, MintPoolEntry mintPoolEntry);
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/wallet.h]: Do you mean passing `mintPoolEntry` by reference?

File: src/test/hdmint_tests.cpp [hunk 2]
```diff
@@ -150,4 +155,93 @@ BOOST_AUTO_TEST_CASE(deterministic)
...
+        return sigma::PrivateCoin(sigmaParams, denom);
+    });
+    vector<CHDMint> vDMints;
+    auto vecSend = CWallet::CreateSigmaMintRecipients(privCoins, vDMints);
+
+    BOOST_CHECK(vDMints[0].GetCount() == TOTAL_MINTS);
+    BOOST_CHECK(zwalletMain->GetCount() == (TOTAL_MINTS+1));
+
+}
+
```
Comment (psolstice) [inline src/test/hdmint_tests.cpp]: I think we need a test where you are given a blockchain already having mints generated with current seed and future values of the counter. Then we need to ensure the code is picking that up correctly and generates next mint with no collision with the ones already in the blockchain. This is the current issue with mint creation after all.

File: src/hdmint/wallet.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/hdmint/wallet.h]: sorry my description isn't clear I mean `const sigma::CoinDenomination& denom` should be just `sigma::CoinDenomination denom` instead.

**Comment by thebevrishot** [comment]
@riordant please resolve remaining request changes.

File: src/test/hdmint_tests.cpp (hunk 2 above)
Comment (riordant) [reply src/test/hdmint_tests.cpp]: Agree. Working on it

**Pull request merged**